File: app/src/main/java/com/playgilround/schedule/client/signin/SignInPresenter.java
package com.playgilround.schedule.client.signin;

import android.content.Context;
import android.content.Intent;

import com.playgilround.schedule.client.ScheduleApplication;
import com.playgilround.schedule.client.data.repository.UsersRepository;
import com.playgilround.schedule.client.data.source.UsersDataSource;

import javax.inject.Inject;

public class SignInPresenter implements SignInContract.Presenter, UsersDataSource.LoginCallBack {

    private static final String TAG = SignInPresenter.class.getSimpleName();

    public static final int ERROR_EMAIL = 0x0001;
    public static final int ERROR_PASSWORD = 0x0002;
    public static final int ERROR_FAIL_SIGN_IN = 0x0003;
    public static final int ERROR_NETWORK_CUSTOM = 0x0004;

    private final Context mContext;
    private final SignInContract.View mView;

    @Inject UsersRepository mRepository;

    SignInPresenter(Context context, SignInContract.View view) {
        mView = view;
        mView.setPresenter(this);
        mContext = context;
        ((ScheduleApplication)(context.getApplicationContext())).appComponent.inject(this);
    }

    @Override
    public void start() {

    }

    @Override
    public boolean checkAutoSignIn() {
        return mRepository.getCurrentUser(mContext) != null;
    }

    @Override
    public void autoSignIn() {
        mView.showProgressBar();
        mRepository.tokenLogin(this);
    }

    @Override
    public void signIn(String email, String password) {
        mView.showProgressBar();
        mRepository.login(email, password, this);
    }

    @Override
    public Intent googleSingIn() {
        return mRepository.googleLogin();
    }

    @Override
    public void firebaseAuthGoogle(Intent data) {
        mRepository.firebaseAuthGoogle(data);
    }

    @Override
    public void onUserLoaded() {
        mView.signInComplete();
    }

    @Override
    public void onDataNotAvailable(int error) {
        mView.signInError(error);
    }
}

File: app/src/main/java/com/playgilround/schedule/client/signin/SignInPresenter.kt
package com.playgilround.schedule.client.signin

import android.app.Activity
import android.content.Context
import android.content.Intent
import com.facebook.CallbackManager
import com.nhn.android.naverlogin.OAuthLogin
import com.playgilround.schedule.client.ScheduleApplication
import com.playgilround.schedule.client.data.repository.UsersRepository
import com.playgilround.schedule.client.data.source.UsersDataSource
import javax.inject.Inject

class SignInPresenter constructor(private val mContext: Context, private val mView: SignInContract.View):
        SignInContract.Presenter, UsersDataSource.LoginCallBack {

    @Inject
    internal lateinit var mRepository: UsersRepository

    init {
        mView.setPresenter(this)
        (mContext.applicationContext as ScheduleApplication).appComponent.signInInject(this)

    }

    override fun start() {

    }

    override fun checkAutoSignIn(): Boolean {
        return mRepository.getCurrentUser(mContext) != null
    }

    override fun autoSignIn() {
        mView.showProgressBar()
        mRepository.tokenLogin(this)
    }

    override fun signIn(email: String, password: String) {
        mView.showProgressBar()
        mRepository.login(email, password, this)
    }

    override fun facebookSignIn(activity: Activity): CallbackManager {
        return mRepository.facebookLogin(activity, this)
    }

    override fun naverInit(): OAuthLogin {
        return mRepository.naverInit()
    }

    override fun naverSignIn(activity: Activity, oAuthLogin: OAuthLogin) {
        mRepository.naverLogin(activity, oAuthLogin, this)
    }

    override fun kakaoSignIn(activity: Activity) {
        mRepository.kakaoLogin(activity, this)
    }

    override fun googleSignIn(): Intent {
        return mRepository.googleLogin()
    }

    override fun firebaseAuthGoogle(data: Intent) {
        mRepository.firebaseAuthGoogle(data, this)
    }

    override fun onUserLoaded() {
        mView.signInComplete()
    }

    override fun onDataNotAvailable(error: Int) {
        mView.signInError(error)
    }

    companion object {
        const val ERROR_EMAIL = 0x0001
        const val ERROR_PASSWORD = 0x0002
        const val ERROR_FAIL_SIGN_IN = 0x0003
        const val ERROR_NETWORK_CUSTOM = 0x0004
    }
}